xmodule/services.py

```diff
@@ -308,3 +312,112 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ResultService(Service):
+    """
+    An XBlock Service that allows XModules to define:
+
+    - if the answers are visible
+    - whether correctness is currently hidden. When correctness is hidden, this
+    limits the user's access to the correct/incorrect flags, messages, problem
+    scores, and aggregate subsection and course grades.
+    """
+    def __init__(self, block=None, user_is_staff=False, **kwargs):
+        super().__init__(**kwargs)
+        self._user_is_staff = user_is_staff
+        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
+        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
+        self._xblock = self.xblock() or block
```

Comment:
@ormsbee, I believe that the this behavior was [implemented](https://github.com/openedx/edx-platform/commit/3ed3fea2fbb14b6c64a406a9226f34b4c2136484#diff-8552615a0e0066a72d15c4de70a3fc92eb10bfd8e148fe566bce2937a7bb52fcR1796-R1799) a long time ago for the `*I18nService`.

It feels somewhat redundant to pass an XBlock instance while calling a service's method because:
1. The base [Service](https://github.com/openedx/XBlock/blob/391f1f552061e19f7936f5bdcd260abc4c59abf3/xblock/reference/plugins.py#L65-L102) class implements a mechanism to retrieve an XBlock. We cannot use it directly, though, because the ways we're passing the instance are [inconsistent](https://github.com/openedx/edx-platform/blob/b453e4861f6cc29dacd6f8c3738009b80cb10c08/openedx/core/djangoapps/xblock/runtime/runtime.py#L282-L289)  across the platform. If we want to keep passing an XBlock instance, it would make sense to unify these approaches to make this less confusing.
2. We must pass the block instance to retrieve the service from the runtime anyway, so we can check the service declarations ("need", "want"). To pass an instance to the method, we need to do `block.runtime(block, "service_name").method(block, *args, **kwargs)`.

If we want to stop using this pattern, we should add comments to the places that use it to indicate the desired direction. That said, I'm okay with either of these approaches. Removing these patterns from the existing services will be more tricky, though, as it would break the compatibility with existing XBlocks.

Reply:
Which approach should I leave in this PR @ormsbee @Agrendalath?

Reply:
Go ahead with @Agrendalath's suggestion. Thank you.